### DISCUSSION TITLE
How to use (OAuth2 with Password (and hashing), Bearer with JWT tokens) with your own Logging page

### DISCUSSION THREAD

**Discussion by SWimplifyDev** [fastapi/fastapi#11891] [Questions] [answered]
### First Check

- [x] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.post("/token")
async def login_for_access_token(
    form_data: Annotated[OAuth2PasswordRequestForm, Depends()],
) -> Token:
    user = authenticate_user(fake_users_db, form_data.username, form_data.password)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
            headers={"WWW-Authenticate": "Bearer"},
        )
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": user.username}, expires_delta=access_token_expires
    )
    return Token(access_token=access_token, token_type="bearer")
```

### Description

There is quite a lot details on FastAPI documentation on how to use [OAuth2 with Password (and hashing), Bearer with JWT tokens]. However when you try to apply those concepts and building your own Login Page for users, things start to look not easy at all. This has been my journey so far:

1. (EASY) Building a /token end point POST, so I can use it on my logging page to retrieve username and password from a user and  obtain a TOKEN as a response if the user exist on my DB

2. (BLACK BOX)The TOKEN is sent to the client or browser. QUESTION: How is this token supposed to be saved on the client? I does not happen behind the scenes.

3. (BLACK BOX) When the client get its TOKEN, he is supposed to be redirected to a protected page. QUESTION: Who is supposed to redirect the user? The server or software running on the users browser?

4. (BLACK BOX) From the very moment the USER gets a TOKEN, all GET request are supposed to be sent with an Authorisation Headers. QUESTION: Software running on the browser should gather the TOKEN and put it on any GET request to the server? How does this happen?

I  appreciate if someone that has already go trough the pain, could help me to solve those BLACK BOXES.

Thanks

### Operating System

macOS

### FastAPI Version

python -c "import fastapi; print(fastapi.__version__)"

### Pydantic Version

python -c "import pydantic; print(pydantic.version.VERSION)"

### Python Version

python --version

**Comment by sinisaos** [comment]
@SWimplifyDev You have to do everything by yourself because the browser does not automatically send tokens in headers. 

> * (BLACK BOX)The TOKEN is sent to the client or browser. QUESTION: How is this token supposed to be saved on the client? I does not happen behind the scenes.

You have several options for saving the token, namely: `local/session storage`, `HttpOnly cookie`, `in browser memory` or a combination of several methods, etc. It's a security issue that has many interpretations. Just google it `stores the jwt token securely in the browser` and it's best to see for yourself which solution you want to apply. [Here](https://github.com/sinisaos/piccolo-vue-starter) is an example with FastAPI + Piccolo ORM as the backend and Vue as the frontend (there are probably a lot of other examples on Github). I don't know if it's the best solution, but on the backend I used [HttpOnly cookie](https://github.com/sinisaos/piccolo-vue-starter/blob/main/backend/accounts/endpoints.py#L83-L91) with [Piccolo CSRFMiddleware](https://github.com/sinisaos/piccolo-vue-starter/blob/dc2ca92d930650350167c354c8440e5066f1dd11/backend/app.py#L84-L85). You must also provide [FastAPI CORSMiddleware](https://github.com/sinisaos/piccolo-vue-starter/blob/dc2ca92d930650350167c354c8440e5066f1dd11/backend/app.py#L34C1-L44C2) to enable frontend and backend communication. I've been using Vue2 in frontend (probably needs a package update) with the [Axios HTTP client](https://axios-http.com/). Here are the Axios [default configuration settings](https://github.com/sinisaos/piccolo-vue-starter/blob/main/frontend/src/main.js#L12C1-L15C50) that will be applied to each request. This configuration is required for proper operation with auth backend.

> * (BLACK BOX) When the client get its TOKEN, he is supposed to be redirected to a protected page. QUESTION: Who is supposed to redirect the user? The server or software running on the users browser?
> * (BLACK BOX) From the very moment the USER gets a TOKEN, all GET request are supposed to be sent with an Authorisation Headers. QUESTION: Software running on the browser should gather the TOKEN and put it on any GET request to the server? How does this happen?

When the frontend receives and stores the token in the cookie, Axios sends a [GET request](https://github.com/sinisaos/piccolo-vue-starter/blob/dc2ca92d930650350167c354c8440e5066f1dd11/frontend/src/store/modules/users.js#L27C3-L30C5) (which has the authorization token in the cookie that we set previously in Axios config defaults) to the [server endpoint](https://github.com/sinisaos/piccolo-vue-starter/blob/dc2ca92d930650350167c354c8440e5066f1dd11/backend/accounts/endpoints.py#L113C1-L117C24) using the [Vue router](https://github.com/sinisaos/piccolo-vue-starter/blob/dc2ca92d930650350167c354c8440e5066f1dd11/frontend/src/router/index.js#L31-L36) (according to the meta tag in route you can see that the route requires auth). Again, I'm not saying this is the best way, but I hope you get an idea of ​​how to do it and I hope you find it useful.

**Reply by SWimplifyDev** [reply]
Thanks @sinisaos  for you response, I thought I was missing something while reading the docs, as the /docs front-end when testing, seems to do the auth stuff like magic without storing anything on the browser. I am trying to explore the most simple way of achieving a login page with fastAPI without needing complex front-end frameworks or any at all. I am scratching my head around how it is possible that PHP could achieve auth so easily on the backend without the need of running any conde on the front end, and yet FastAPI being so powerful and complete could not do that as simple as PHP.

**Reply by sinisaos** [reply]
@SWimplifyDev Sorry, I thought you needed auth with frontend framework which is much more complicated. If you need server side auth, it is much simpler and you can use session auth with cookies similar to Django or PHP etc. [Here is an example](https://github.com/sinisaos/starlette-piccolo-rental) of Starlette session auth with Piccolo ORM. If you dont use SQLAlchemy and want to use Piccolo ORM, Piccolo has [SessionAuth](https://piccolo-api.readthedocs.io/en/latest/session_auth/index.html#sessionauth) middleware with login/register/logout endpoints and server side templates out of a box.

**Comment by SWimplifyDev** [comment]
@sinisaos, thanks for sharing, I will have a look.

**Comment by SWimplifyDev** [comment]
@sinisaos I would like very much to hear your thoughts on this minimalist approach for authentication, it works quite nice:

![Screenshot 2024-07-29 at 4 14 16 PM](https://github.com/user-attachments/assets/6215de57-153b-459b-ba6d-19408a91df96)

**Reply by sinisaos** [reply]
@SWimplifyDev It looks good to me for the backend part and is very similar to the example given in the FastAPI documentation or the example from my repositories that use cookie based auth. But maybe I didn't understand your question well. I thought you needed the frontend part (login page in frontend JS framework or server side rendering HTML template). I don't see that in your example. If you use cookies, you must provide some CSRF middleware to protect from CSRF attacks.

**Comment by SWimplifyDev** [comment]
@sinisaos , I do have that part to render HTML for my login page, I just did not want to make my post that long:
![Screenshot 2024-07-29 at 5 33 04 PM](https://github.com/user-attachments/assets/ca194545-0a14-4ac4-94c5-d07965af849a)

How could I add that CSRF middleware, could you guide me to find some documentation on that part pls

**Reply by sinisaos** [reply]
@SWimplifyDev I always use Piccolo as an ORM, and in Piccolo we have a [CSRF middleware](https://piccolo-api.readthedocs.io/en/latest/csrf/index.html) already ready to go. If you are using a different ORM, you can check out a different Starlette middleware that you can find online. I used [this library](https://github.com/simonw/asgi-csrf) a long time ago and it did the job. [Here](https://github.com/sinisaos/starlette-piccolo-orm/blob/master/app.py#L93-L94) and [here](https://github.com/sinisaos/starlette-piccolo-orm/blob/f87cc867d951fcef5931f0eef2118fe3e47328fa/templates/accounts/login.html#L18) is an example of usage or you can read library `readme.md`.

**Comment by SWimplifyDev** [comment]
@sinisaos, I am using very lite DB document base called tinyDB. This is my best approach to get a simple, functional Auth flow for any SaaS project [FastAuthBase](https://github.com/SWimplifyDev/FastAuthBase). I am not sure what lever of security I am now with this approach as I can't find a simple way to implement the CSRF middleware you mentioned above.
![Screenshot 2024-07-31 at 10 33 25 AM](https://github.com/user-attachments/assets/4b603276-195c-493d-a101-8653bb439b63)
![Screenshot 2024-07-31 at 10 33 40 AM](https://github.com/user-attachments/assets/d61711cd-2184-49bb-9097-68214061cd61)
![Screenshot 2024-07-31 at 10 33 58 AM](https://github.com/user-attachments/assets/a5fedd7b-ea3f-4c47-96af-66321c425036)

In your opinion, how critical would be not to have the CSRF midd if I am using cookies, I want to keep this approach as lean as possible so I can share with people that are just staring with FastAPI.

I appreciate your thoughts and knowledge, thanks!

**Reply by sinisaos** [reply, marked as answer]
@SWimplifyDev [asgi-csrf](https://github.com/simonw/asgi-csrf) library works fine for me, but as stated in the `readme.md` **"Your application code should include that csrftoken as a hidden form field in all POST forms:"**. I also changed a user logout. First render a logout template, then in that template make POST form (with csrftoken included) to log the user out. Here's a video showing how it all works. 

[csfr.webm](https://github.com/user-attachments/assets/f0a10bdb-419f-4f3b-8bb0-04a233d5e673)

I will send a PR request to your repo with the necessary changes (and some linting) so you can see the changes..

**Reply by SWimplifyDev** [reply]
What an elegant solution @sinisaos. I can't thank you enough for helping me on this one. I really appreciate it.

**Reply by sinisaos** [reply]
@SWimplifyDev No problem, I'm happy to help. I made another PR to use environment variables (the secret key should be secret). After that I leave you to structure your project as you wish.
